Engineering/Mobile/ios_bluetooth_low_energy/ios_bluetooth_low_energy.md

```diff
@@ -0,0 +1,376 @@
+                                                      properties: .indicate,
+                                                      value: nil,
+                                                      permissions: .readable)
+        self.charForIndicate = charForIndicate
+
+        // create service
+        let service = CBMutableService(type: uuidService, primary: true)
+        service.characteristics = [charForRead, charForWrite, charForIndicate]
+        return service
+    }
+```
+
+* Handle data reviece from BLE Central
+
+```
+ private func handleDataResponse(_ dataResponse: Data) {
+        var textJson = String(data: dataResponse, encoding: .utf8) ?? ""
+        if let indexRemove = textJson.firstIndex(where: { $0 == "<" }) {
+            textJson.remove(at: indexRemove)
+        }
+        if let indexRemove = textJson.firstIndex(where: { $0 == ">" }) {
+            textJson.remove(at: indexRemove)
+        }
+        convertToJson(with: textJson)
+    }
+    
+    private func convertToJson(with text: String) {
+        guard let data = text.data(using: .utf8),
+              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
+              let pass = json["password"] as? String
+        else {
+            sendNotify(isResponseSuccess: false)
+            return
+        }
+        sendNotify(isResponseSuccess: pass == "1")
+    }
+    
+    private func sendNotify(isResponseSuccess: Bool) {
+        let textSuccess = "<{\"action\":\"bleconfig\",\"result\":\"ok\"}>"
+        let textFail = "<{\"action\":\"bleconfig\",\"result\":\"ng\"}>"
+        bleSendIndication(isResponseSuccess ? textSuccess : textFail)
+    }
+```
+
+* Function didReceiveWrite will receive data from BLE Central
+
+```
+    func peripheralManager(_ peripheral: CBPeripheralManager, didReceiveWrite requests: [CBATTRequest]) {
+        var log = "didReceiveWrite requests.count = \(requests.count)"
+        requests.forEach { (request) in
+            log += "\nrequest.offset: \(request.offset)"
+            log += "\nrequest.char.UUID: \(request.characteristic.uuid.uuidString)"
+            switch request.characteristic.uuid {
+            case uuidCharForWrite:
+                if let data = request.value {
+                    handleDataResponse(data)
+                    let textValue = String(data: data, encoding: .utf8)
+                    log += "\nresponding with success, value = '\(String(describing: textValue))'"
+                    blePeripheral.respond(to: request, withResult: .success)
+                }
+            default:
+                log += "\nresponding with attributeNotFound"
+                blePeripheral.respond(to: request, withResult: .attributeNotFound)
+            }
+        }
+        appendLog(log)
+    }
+```
+
+* When receive data from BLE Central, BLE Prepheral will notify success or fail to BLE Central
+
+```
+    private func bleSendIndication(_ valueString: String) {
```

Comment:
Same here, can you remove the leading space and specify the language for the block?